src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -1779,7 +1780,32 @@ void StatefulWriter::send_heartbeat_nts_(
     {
         assert(firstSeq <= lastSeq);
 
-        // Check if it has to be sent a GAP with the gaps in the history
+        if (!liveliness)
+        {
+            if (SequenceNumber_t::unknown() != firstSeq &&
+                    lastSeq.to64long() - firstSeq.to64long() + 1 != mp_history->getHistorySize())
+            {
+                RTPSGapBuilder gaps(message_group);
+
+                // There are holes in the history.
+                History::const_iterator cit = mp_history->changesBegin();
+                SequenceNumber_t prev = (*cit)->sequenceNumber + 1;
+                ++cit;
+                while (cit != mp_history->changesEnd())
+                {
+                    while (prev != (*cit)->sequenceNumber)
+                    {
+                        gaps.add(prev);
+                        ++prev;
+                    }
+
+                    ++prev;
+                    ++cit;
+                }
+
+                gaps.flush();
+            }
+        }
```

Comment:
This code belongs in `send_heartbeat_to_nts`, as it is the one used for separate sending.

Method `send_heartbeat_to_nts_` is also called inside `send_heartbeat_piggyback_nts_`, and `send_heartbeat_to_all_readers`, just after adding gaps to the message group.

Reply:
Additionally, since this seems to have been copied from line 607, I suggest we add a new private method `add_gaps_for_holes_in_history_(RTPSMessageGroup& group)`